docs/spec/relayer/ICS18Environment.tla

```diff
@@ -0,0 +1,314 @@
+          chainBstore, \* store of ChainB
+          incomingDatagramsChainA, \* set of datagrams incoming to ChainA
+          incomingDatagramsChainB, \* set of datagrams incoming to ChainB
+          relayer1Heights, \* the client heights of Relayer1
+          relayer2Heights, \* the client heights of Relayer2
+          outgoingDatagrams \* sets of datagrams outgoing of the relayers
+          
+vars == <<chainAstore, chainBstore, 
+          incomingDatagramsChainA, incomingDatagramsChainB,
+          relayer1Heights, relayer2Heights,
+          outgoingDatagrams>>
+chainAvars == <<chainAstore, incomingDatagramsChainA>>
+chainBvars == <<chainBstore, incomingDatagramsChainB>>
+relayerVars == <<relayer1Heights, relayer2Heights, outgoingDatagrams>>
+
+Heights == 1..MaxHeight \* set of possible heights of the chains in the system
+
+(***************************************************************************
+ Instances of Relayer and Chain
+ ***************************************************************************)
+
+\* We suppose there are two correct relayers in the system, Relayer1 and Relayer2
+\* Relayer1 -- Instance of Relayer.tla
+Relayer1 == INSTANCE Relayer
+            WITH GenerateClientDatagrams <- ClientDatagramsRelayer1,
+                 GenerateConnectionDatagrams <- ConnectionDatagramsRelayer1,
+                 GenerateChannelDatagrams <- ChannelDatagramsRelayer1,
+                 Heights <- Heights,
+                 chainAstore <- chainAstore,
+                 chainBstore <- chainBstore,
+                 relayerHeights <- relayer1Heights
+                 
+\* Relayer2 -- Instance of Relayer.tla      
+Relayer2 == INSTANCE Relayer
+            WITH GenerateClientDatagrams <- ClientDatagramsRelayer2,
+                 GenerateConnectionDatagrams <- ConnectionDatagramsRelayer2,
+                 GenerateChannelDatagrams <- ChannelDatagramsRelayer2,
+                 Heights <- Heights,
+                 chainAstore <- chainAstore,
+                 chainBstore <- chainBstore,
+                 relayerHeights <- relayer2Heights
+
+\* We suppose there are two chains that communicate, ChainA and ChainB
+\* ChainA -- Instance of Chain.tla
+ChainA == INSTANCE Chain
+          WITH ChainID <- "chainA",
+               Heights <- Heights,
+               chainStore <- chainAstore,
+               incomingDatagrams <- incomingDatagramsChainA
+
+\* ChainB -- Instance of Chain.tla 
+ChainB == INSTANCE Chain
+          WITH ChainID <- "chainB",
+               Heights <- Heights,
+               chainStore <- chainBstore,
+               incomingDatagrams <- incomingDatagramsChainB
+
+(***************************************************************************
+ Component actions
+ ***************************************************************************)
+
+\* RelayerAction: either correct relayer takes a step, leaving the other 
+\* variables unchanged 
+RelayerAction ==
+    \/ /\ Relayer1!Next
```

Comment:
We will probably need to make relayer steps more fine grained, i.e., smaller atomic steps. In this version a relayer comes, query state of both chains, create datagrams and submit them, and only then the second relayer can do something. This constrains unrealistically system concurrency. But that is probably out of scope of this PR.

Reply:
It is a bit more fine-grained than that. Currently, in a relayer step, defined by Relayer!Next (lines 249-252 of Relayer.tla), we have that a relayer either increases the client height of one chain (action `UpdateRelayerClients(chainID)`, lines 206-212 of Relayer.tla), or it scans srcChain and creates datagrams for dstChain (action `Relay(srcChainID, dstChainID)`, lines 216-226 of Relayer.tla). These datagrams are only created, not delivered. The environment delivers them using `SubmitDatagrams`, adding more concurrency. I assume we can go even more fine-grained if we define separate actions for creating datagrams of each protocol (client, connection, channel...), as well as submitting datagrams one by one, rather than in a batch.

I agree that the current model does not capture the concurrency in the real world, but I think we should be careful when adding more behaviors, in order not to make model checking infeasible.